Codegen/src/Generators/DefaultGenerator.cs

```diff
@@ -638,11 +688,245 @@ internal void UpdateDeltas(SourceText targetSourceText)
+
+                if (environmentDiv != null)
+                {
+                    foreach (ITextLine environmentDivLine in environmentDiv.SelfAndChildrenLines)
+                    {
+                        sw.WriteLine(environmentDivLine.Text); GSLineOffset += 1;
+                    }
+                }
+            }
+
+            sw.WriteLine("       DATA DIVISION."); GSLineOffset += 1;
+            sw.WriteLine("       WORKING-STORAGE SECTION."); GSLineOffset += 1;
+            sw.WriteLine("       01 PIC X(8) value ':TC:GBLS'."); GSLineOffset += 1;
+
+            sw.WriteLine(GetGlobalStorageData(clonedMapper, Input, lmCtx));
+            //So Here we can relocate Global Storage Line Mapping
+            if (lmCtx != null)
+            {
+                for (int i = FirstGSLine; i < LastGSLine; i++)
+                {
+                    int firstLine = GSLmCtx.LineMapping[i].Item1;
+                    int lastLine = GSLmCtx.LineMapping[i].Item2;
+                    lmCtx.LineMapping[i] = new Tuple<int, int>(firstLine + lmCtx.startLineMapCounter + GSLineOffset - 1, lastLine + lmCtx.startLineMapCounter + GSLineOffset - 1);
+                }
+
+                //Create also the inverse line mapping for Global Storage generated in nested programs and procedures
+                if (lmCtx.AllGlobalStorageLineDelta != null)
+                {
+                    lmCtx.InverseLineMapping = new List<Tuple<int, int>>();
+                    foreach (int gsDelta in lmCtx.AllGlobalStorageLineDelta)
+                    {
+                        int baseLine = GSLmCtx.LineMapping[FirstGSLine].Item1 + lmCtx.startLineMapCounter + GSLineOffset - 1;
+                        for (int i = FirstGSLine; i < LastGSLine; i++)
+                        {
+                            int firstLine = GSLmCtx.LineMapping[i].Item1 + lmCtx.startLineMapCounter + GSLineOffset - 1;
+                            int lastLine = GSLmCtx.LineMapping[i].Item2 + lmCtx.startLineMapCounter + GSLineOffset - 1;
+                            for (int j = firstLine; j <= lastLine; j++)
+                            {
+                                Tuple<int, int> inv = new Tuple<int, int>(-(j - baseLine + gsDelta + GSLineOffsetInverse), i + 1);
+                                lmCtx.InverseLineMapping.Add(inv);
+                            }
+                        }
+                    }
+                }
+            }
+
+            sw.WriteLine( "       LINKAGE SECTION.");
+            sw.WriteLine( "       01 GlobalPointer pointer.");
+            sw.WriteLine( "       PROCEDURE DIVISION USING BY REFERENCE GlobalPointer.");
+            sw.WriteLine( "           set GlobalPointer to address of TC-GlobalData");
+            sw.WriteLine( "           .");
+            sw.WriteLine($"       END PROGRAM {RootNode.MainProgram.Hash}.");
+
+            return sw.ToString();
+        }
+
+        public string GlobalStorageData = null;
```

Comment:
Public field is bad design as it can be modified by anything that accesses the class. It is currently required by the GlobalStorage contribute action but should be reworked.